File: app/src/main/java/com/geridea/trentastico/gui/views/CourseTimesCalendar.java
package com.geridea.trentastico.gui.views;


/*
 * Created with ♥ by Slava on 11/03/2017.
 */

import android.content.Context;
import android.support.annotation.NonNull;
import android.util.AttributeSet;

import com.alamkanak.weekview.DateTimeInterpreter;
import com.alamkanak.weekview.WeekViewEvent;
import com.android.volley.VolleyError;
import com.geridea.trentastico.model.LessonSchedule;
import com.geridea.trentastico.model.LessonType;
import com.geridea.trentastico.model.LessonsSet;
import com.geridea.trentastico.model.cache.CachedLessonsSet;
import com.geridea.trentastico.network.LessonsLoader;
import com.geridea.trentastico.network.operations.ILoadingOperation;
import com.geridea.trentastico.network.operations.ParsingErrorOperation;
import com.geridea.trentastico.network.operations.ReadingErrorOperation;
import com.geridea.trentastico.utils.time.CalendarUtils;
import com.geridea.trentastico.utils.time.WeekInterval;
import com.threerings.signals.Listener1;
import com.threerings.signals.Listener2;
import com.threerings.signals.Signal0;
import com.threerings.signals.Signal1;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collection;
import java.util.List;
import java.util.Locale;

import static com.geridea.trentastico.Config.DEBUG_MODE;

public class CourseTimesCalendar extends CustomWeekView implements CustomWeekView.ScrollListener {

    private final static SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("EEEE d MMMM", Locale.ITALIAN);
    private final static SimpleDateFormat DATE_FORMAT_DEBUG = new SimpleDateFormat("(w) EEEE d MMMM", Locale.ITALIAN);

    private final static SimpleDateFormat DATE_FORMAT_MEDIUM = new SimpleDateFormat("EE dd/MM", Locale.ITALIAN);
    private final static SimpleDateFormat DATE_FORMAT_MEDIUM_DEBUG = new SimpleDateFormat("(w)EE dd/MM", Locale.ITALIAN);

    private final static SimpleDateFormat DATE_FORMAT_SHORT  = new SimpleDateFormat("dd/MM", Locale.ITALIAN);
    private final static SimpleDateFormat DATE_FORMAT_SHORT_DEBUG  = new SimpleDateFormat("(w)dd/MM", Locale.ITALIAN);

    private final static SimpleDateFormat DATE_FORMAT_SHORT_ONLY_DAY = new SimpleDateFormat("EE", Locale.ITALIAN);

    private final static SimpleDateFormat FORMAT_ONLY_DAY = new SimpleDateFormat("EEEE", Locale.ITALIAN);

    //Signals
    /**
     * Dispatched when the calendar is about to perform a loading from the network.<br>
     * WARNING: may be called on a not-UI thread!
     */
    public final Signal1<ILoadingOperation> onLoadingOperationResult = new Signal1<>();

    /**
     * Dispatched when the calendar has finished loading something from the network.<br>
     * WARNING: may be called on a not-UI thread!
     */
    public final Signal0 onLoadingOperationFinished = new Signal0();

    //Data
    private final LessonsSet currentlyShownLessonsSet = new LessonsSet();
    private LessonsLoader loader;

    public CourseTimesCalendar(Context context) {
        super(context);
        initCalendar();
    }

    public CourseTimesCalendar(Context context, AttributeSet attrs) {
        super(context, attrs);
        initCalendar();
    }

    public CourseTimesCalendar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initCalendar();
    }

    private void initCalendar() {
        prepareLoader();
        setDateTimeInterpreter(getAppropriateDateTimeInterpreter());
        setScrollListener(this);
    }

    private DateTimeInterpreter getAppropriateDateTimeInterpreter() {
        return getDateInterpreterForNumberOfDays(getNumberOfVisibleDays());
    }

    private DateTimeInterpreter getDateInterpreterForNumberOfDays(int numberOfVisibleDays) {
        if(numberOfVisibleDays <= 2){
            return new DateTimeInterpreter() {
                @Override
                public String interpretDate(Calendar date) {
                    Calendar today = Calendar.getInstance();
                    if (isSameDay(today, date)) {
                        return "Oggi (" + FORMAT_ONLY_DAY.format(date.getTime()) + ")";
                    }

                    today.add(Calendar.DAY_OF_MONTH, +1);
                    if (isSameDay(today, date)) {
                        return "Domani (" + FORMAT_ONLY_DAY.format(date.getTime()) + ")";
                    }

                    today.add(Calendar.DAY_OF_MONTH, +1);
                    if (isSameDay(today, date)) {
                        return "Dopodomani (" + FORMAT_ONLY_DAY.format(date.getTime()) + ")";
                    }

                    today.add(Calendar.DAY_OF_MONTH, -3);
                    if (isSameDay(today, date)) {
                        return "Ieri (" + FORMAT_ONLY_DAY.format(date.getTime()) + ")";
                    }

                    return (DEBUG_MODE ? DATE_FORMAT_DEBUG : DATE_FORMAT).format(date.getTime());
                }

                @Override
                public String interpretTime(int hour) {
                    return interpretHours(hour);
                }
            };
        } else if(numberOfVisibleDays <= 5) {
            return new DateTimeInterpreter() {
                @Override
                public String interpretDate(Calendar date) {
                    Calendar today = Calendar.getInstance();
                    if (isSameDay(today, date)) {
                        return "Oggi";
                    }

                    today.add(Calendar.DAY_OF_MONTH, +1);
                    if (isSameDay(today, date)) {
                        return "Domani";
                    }

                    today.add(Calendar.DAY_OF_MONTH, +1);
                    if (isSameDay(today, date)) {
                        return "Dopodomani";
                    }

                    today.add(Calendar.DAY_OF_MONTH, -3);
                    if (isSameDay(today, date)) {
                        return "Ieri";
                    }

                    return (DEBUG_MODE ? DATE_FORMAT_MEDIUM_DEBUG : DATE_FORMAT_MEDIUM)
                                .format(date.getTime());
                }

                @Override
                public String interpretTime(int hour) {
                    return interpretHours(hour);
                }
            };
        } else {// > 6
            return new DateTimeInterpreter() {
                @Override
                public String interpretDate(Calendar date) {
                    Calendar firstDay = CalendarUtils.calculateFirstDayOfWeek();
                    Calendar lastDay  = (Calendar) firstDay.clone();
                    lastDay.add(Calendar.WEEK_OF_MONTH, +1);

                    if (date.equals(firstDay) || (date.after(firstDay) && date.before(lastDay))) {
                        return (DEBUG_MODE ? DATE_FORMAT_SHORT_DEBUG : DATE_FORMAT_SHORT_ONLY_DAY)
                                .format(date.getTime());
                    } else {
                        return (DEBUG_MODE ? DATE_FORMAT_SHORT_DEBUG : DATE_FORMAT_SHORT)
                                    .format(date.getTime());
                    }
                }

                @Override
                public String interpretTime(int hour) {
                    return interpretHours(hour);
                }
            };
        }

    }

    @NonNull
    private String interpretHours(int hour) {
        return hour + ":00";
    }

    private void prepareLoader() {
        loader = new LessonsLoader();
        loader.onLoadingOperationStarted.connect(new Listener1<ILoadingOperation>() {
            @Override
            public void apply(ILoadingOperation operation) {
                onLoadingOperationResult.dispatch(operation);
            }
        });
        loader.onLoadingOperationSuccessful.connect(new Listener2<LessonsSet, WeekInterval>() {
            @Override
            public void apply(LessonsSet lessons, WeekInterval interval) {
                currentlyShownLessonsSet.mergeWith(lessons);

                addEnabledInterval(interval);

                addEventsFromLessonsSet(lessons);

                onLoadingOperationFinished.dispatch();
            }
        });

        loader.onPartiallyCachedResultsFetched.connect(new Listener1<CachedLessonsSet>() {
            @Override
            public void apply(CachedLessonsSet lessonsSet) {
                currentlyShownLessonsSet.mergeWith(lessonsSet);

                ArrayList<WeekInterval> cachedIntervals = lessonsSet.getCachedIntervals();
                addEnabledIntervals(cachedIntervals);

                addEventsFromLessonsSet(lessonsSet);
            }
        });

        loader.onLoadingErrorHappened.connect(new Listener1<VolleyError>() {
            @Override
            public void apply(VolleyError error) {
                onLoadingOperationResult.dispatch(new ReadingErrorOperation());
            }
        });
        loader.onParsingErrorHappened.connect(new Listener1<Exception>() {
            @Override
            public void apply(Exception e) {
                e.printStackTrace();
                onLoadingOperationResult.dispatch(new ParsingErrorOperation());
            }
        });
    }

    private void addEnabledIntervals(ArrayList<WeekInterval> intervals) {
        for (WeekInterval interval : intervals) {
            addEnabledInterval(interval);
        }
    }

    public void loadNearEvents() {
        loader.loadEventsNearDay(getFirstVisibleDay());
    }

    private void addEventsFromLessonsSet(LessonsSet lessons) {
        addEvents(makeEventsFromLessonsSet(lessons));
    }

    private boolean isSameDay(Calendar date1, Calendar date2) {
        return date1.get(Calendar.YEAR)         == date2.get(Calendar.YEAR) &&
               date1.get(Calendar.MONTH)        == date2.get(Calendar.MONTH) &&
               date1.get(Calendar.DAY_OF_MONTH) == date2.get(Calendar.DAY_OF_MONTH);
    }

    private List<? extends WeekViewEvent> makeEventsFromLessonsSet(LessonsSet lessonsSet) {
        ArrayList<LessonToEventAdapter> events = new ArrayList<>();
        for (LessonSchedule lessonSchedule : lessonsSet.getScheduledLessons()) {
            events.add(new LessonToEventAdapter(lessonSchedule));
        }

        return events;
    }

    @Override
    public void onFirstVisibleDayChanged(Calendar newFirst, Calendar oldFirst) {
        if (isInEditMode()) {
            return;
        }

        ScrollDirection scrollDirection = newFirst.after(oldFirst) ? ScrollDirection.RIGHT : ScrollDirection.LEFT;

        CustomWeekView.DisabledDayVisibleResult disabledDays = getDisabledDaysVisibleFromDay(newFirst);
        if(disabledDays.isFirstDayDisabled()) {
            loader.loadDaysIfNeeded(disabledDays.getFirstVisibleWeek(), scrollDirection);
        }

        if(disabledDays.isLastDayDisabled()) {
            loader.loadDaysIfNeeded(disabledDays.getLastVisibleWeek(), scrollDirection);
        }
    }

    public Collection<LessonType> getCurrentLessonTypes() {
        return currentlyShownLessonsSet.getLessonTypes();
    }

    public void notifyLessonTypeVisibilityChanged() {
        clear();

        loader.loadEventsNearDay(getFirstVisibleDay());
    }

    public int rotateNumOfDaysShown() {
        int numOfDaysToShow = getNextNumberOfDaysToShow(getNumberOfVisibleDays());
        prepareForNumberOfVisibleDays(numOfDaysToShow);

        return numOfDaysToShow;
    }

    public void prepareForNumberOfVisibleDays(int numOfDaysToShow) {
        setNumberOfVisibleDays(numOfDaysToShow);
        setDateTimeInterpreter(getAppropriateDateTimeInterpreter());
        invalidate();
    }

    /**
     * @deprecated do not call this method directly, but use
     * {@link CourseTimesCalendar#prepareForNumberOfVisibleDays(int)}
     */
    @Override
    @Deprecated
    public void setNumberOfVisibleDays(int numberOfVisibleDays) {
        super.setNumberOfVisibleDays(numberOfVisibleDays);
    }

    private int getNextNumberOfDaysToShow(int numberOfVisibleDays) {
        switch(numberOfVisibleDays){
            case 1: return 2;
            default:
            case 2: return 3;
            case 3: return 7;
            case 7: return 1;
        }
    }

    public static class LessonToEventAdapter extends WeekViewEvent {
        private LessonSchedule lesson;

        public LessonToEventAdapter(LessonSchedule lesson) {
            super(lesson.getId(), lesson.getFullDescription(), lesson.getStartCal(), lesson.getEndCal());
            setColor(lesson.getColor());

            this.lesson = lesson;
        }

        public LessonSchedule getLesson() {
            return lesson;
        }
    }

}

File: app/src/main/java/com/geridea/trentastico/gui/views/CourseTimesCalendar.kt
package com.geridea.trentastico.gui.views


/*
 * Created with ♥ by Slava on 11/03/2017.
 */

import android.content.Context
import android.graphics.RectF
import android.util.AttributeSet
import com.alamkanak.weekview.DateTimeInterpreter
import com.alamkanak.weekview.WeekViewEvent
import com.geridea.trentastico.gui.views.requestloader.ILoadingMessage
import com.geridea.trentastico.model.LessonSchedule
import com.geridea.trentastico.model.LessonType
import com.geridea.trentastico.network.LessonsLoader
import com.geridea.trentastico.utils.ColorDispenser
import com.geridea.trentastico.utils.UIUtils
import com.geridea.trentastico.utils.time.CalendarUtils
import com.geridea.trentastico.utils.time.DayInterval
import com.threerings.signals.Signal1
import com.threerings.signals.Signal2
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList

class CourseTimesCalendar : CustomWeekView, CustomWeekView.EventClickListener {

    //Signals
    /**
     * Dispatched when the user clicks a specific event
     */
    val onEventClicked = Signal2<LessonSchedule, LessonType>()

    /**
     * Dispatched when something worth of note is happening in the calendar. For instance starting
     * or finishing fetching lessons from network.<br></br>
     * WARNING: may be called on a not-UI thread!
     */
    val onLoadingOperationNotify = Signal1<ILoadingMessage>()

    val currentLessonTypes: MutableSet<LessonType> = mutableSetOf()

    //Data
    private lateinit var loader: LessonsLoader

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    constructor(
            context: Context,
            attrs: AttributeSet,
            defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    init {
        prepareLoader()
        dateTimeInterpreter = getDateInterpreterForNumberOfDays(numberOfVisibleDays)
        eventClickListener = this
    }

    private fun getDateInterpreterForNumberOfDays(numberOfVisibleDays: Int): DateTimeInterpreter =
        when {
            numberOfVisibleDays <= 2 -> object : DateTimeInterpreter {
                override fun interpretDate(date: Calendar): String {
                    val formattedDay = FORMAT_ONLY_DAY.format(date.time).toUpperCase()

                    val day = CalendarUtils.debuggableToday
                    if (isSameDay(day, date)) return "$formattedDay\nOggi"

                    day.add(Calendar.DAY_OF_MONTH, +1)
                    return if (isSameDay(day, date)) "$formattedDay\nDomani"
                    else formattedDay+"\n"+ DAY_AND_MONTH_FORMAT.format(date.time)
                }

                override fun interpretTime(hour: Int): String = interpretHours(hour)
            }

            numberOfVisibleDays < 5 -> object : DateTimeInterpreter {
                override fun interpretDate(date: Calendar): String {
                    val dayName = FORMAT_ONLY_DAY.format(date.time).toUpperCase()

                    val day = CalendarUtils.debuggableToday
                    if (isSameDay(day, date)) return "$dayName\nOggi"

                    day.add(Calendar.DAY_OF_MONTH, +1)
                    if (isSameDay(day, date)) return "$dayName\nDomani"

                    return dayName+"\n"+DAY_AND_MONTH_FORMAT.format(date.time)
                }

                override fun interpretTime(hour: Int): String = interpretHours(hour)
            }
            else -> // > 6
                object : DateTimeInterpreter {
                    override fun interpretDate(date: Calendar): String {
                        val dayName     = DATE_FORMAT_SHORT_ONLY_DAY.format(date.time).toUpperCase()
                        val dayAndMonth = DAY_AND_MONTH_SHORT_FORMAT.format(date.time)

                        return dayName+"\n"+dayAndMonth
                    }

                    override fun interpretTime(hour: Int): String = interpretHours(hour)
                }
        }


    private fun interpretHours(hour: Int): String = hour.toString()

    private fun prepareLoader() {
        loader = LessonsLoader()
        loader.onLoadingMessageDispatched  .connect { message -> onLoadingOperationNotify.dispatch(message) }
        loader.onLoadingOperationSuccessful.connect { scheduledLessons, lessonTypes, message ->

            //We cannot use post() here since it's possible that the calendar is still not
            //attached to the fragment
            UIUtils.runOnMainThread {
                currentLessonTypes.addAll(lessonTypes)

                if (scheduledLessons.isNotEmpty()) {
                    //If empty, the covered interval calculation will fail
                    addEnabledInterval(calculateCoveredInterval(scheduledLessons))
                    addLessons(scheduledLessons)

                    if(!loader.hasAnythingBeingLoaded){
                        goToFirstEnabledDayIfNeeded()
                    }
                }

                onLoadingOperationNotify.dispatch(message)
            }
        }
    }

    private fun goToFirstEnabledDayIfNeeded() {
        //We send the user to the first enabled day only in case the lessons of the user still
        // didn't start
        val firstEnabledDay = calculateFirstEnabledDayMillis()
        if(CalendarUtils.debuggableToday.timeInMillis < firstEnabledDay.millis){
            val calendar = CalendarUtils.clearCalendar
            calendar.timeInMillis = firstEnabledDay.millis
            goToDate(calendar)
        }
    }

    /**
     * @return the [DayInterval] that starts with the earliest lesson and ends with the latest one
     */
    private fun calculateCoveredInterval(scheduledLessons: List<LessonSchedule>): DayInterval {
        val minTime = scheduledLessons.minOfOrNull { it.startsAt }
        val maxTime = scheduledLessons.maxOfOrNull { it.endsAt }

        return DayInterval(
            CalendarUtils.getCalendarWithMillis(minTime!!),
            CalendarUtils.getCalendarWithMillis(maxTime!!)
        )
    }

    fun loadEvents() = loader.loadAndAddLessons()

    private fun addLessons(lessons: List<LessonSchedule>) =
            addEvents(makeEventsFromLessons(lessons))

    private fun isSameDay(date1: Calendar, date2: Calendar): Boolean =
            date1.get(Calendar.YEAR) == date2.get(Calendar.YEAR) &&
                    date1.get(Calendar.MONTH) == date2.get(Calendar.MONTH) &&
                    date1.get(Calendar.DAY_OF_MONTH) == date2.get(Calendar.DAY_OF_MONTH)

    private fun makeEventsFromLessons(lessons: List<LessonSchedule>): List<WeekViewEvent> {
        return lessons.mapTo(ArrayList()) {
            LessonToEventAdapter(it, ColorDispenser.getColor(it.lessonTypeId))
        }
    }

    fun notifyEventsChanged() {
        clear()

        loader.loadAndAddLessons()
    }

    fun rotateNumOfDaysShown(): Int {
        val numOfDaysToShow = getNextNumberOfDaysToShow(numberOfVisibleDays)
        prepareForNumberOfVisibleDays(numOfDaysToShow)

        return numOfDaysToShow
    }

    fun prepareForNumberOfVisibleDays(numOfDaysToShow: Int, jumpToFirstVisibleDay: Boolean = true) {
        numberOfVisibleDays = numOfDaysToShow
        dateTimeInterpreter = getDateInterpreterForNumberOfDays(numberOfVisibleDays)

        if(jumpToFirstVisibleDay){
            goToDate(firstVisibleDay) //Fixes #47
        }
        invalidate()
    }

    /**
     * Note: do not call this method directly, but use
     * [CourseTimesCalendar.prepareForNumberOfVisibleDays] instead
     */
    override var numberOfVisibleDays: Int
        get() = super.numberOfVisibleDays
        set(numberOfVisibleDays) {
            super.numberOfVisibleDays = numberOfVisibleDays
        }

    private fun getNextNumberOfDaysToShow(numberOfVisibleDays: Int): Int =
        when (numberOfVisibleDays) {
            1 -> 2
            3 -> 5
            5 -> 1
        /*2*/ else -> 3
        }

    override fun onEventClick(clickedEvent: LessonToEventAdapter, eventRect: RectF?) {
        onEventClicked.dispatch(clickedEvent.lesson, getLessonTypeOfEvent(clickedEvent))
    }

    private fun getLessonTypeOfEvent(event: LessonToEventAdapter) =
            currentLessonTypes.first { it.id == event.lesson.lessonTypeId }

    class LessonToEventAdapter(val lesson: LessonSchedule, color: Int) : WeekViewEvent(
            nextId, lesson.eventDescription, lesson.startCal, lesson.endCal
    ) {

        init {
            this.color = color
        }

        companion object {
            var nextId = 1L
                get() = field++
                private set
        }

    }

    companion object {
        private val DAY_AND_MONTH_FORMAT       = SimpleDateFormat("d MMMM", Locale.ITALIAN)
        private val DAY_AND_MONTH_SHORT_FORMAT = SimpleDateFormat("d MMM",  Locale.ITALIAN)
        private val DATE_FORMAT_SHORT_ONLY_DAY = SimpleDateFormat("EE",     Locale.ITALIAN)
        private val FORMAT_ONLY_DAY            = SimpleDateFormat("EEEE",   Locale.ITALIAN)
    }

}
